source/server/http_test_server_filter.cc

```diff
@@ -16,6 +17,22 @@ HttpTestServerDecoderFilterConfig::HttpTestServerDecoderFilterConfig(
     nighthawk::server::ResponseOptions proto_config)
     : server_config_(std::move(proto_config)) {}
 
+uint64_t HttpTestServerDecoderFilterConfig::ThreadSafeMontonicTimeStopwatch::getElapsedNsAndReset(
+    Envoy::TimeSource& time_source) {
+  Envoy::Thread::LockGuard guard(lock_);
+  // Note that we obtain monotonic time under lock, to ensure that start_ will be updated
+  // monotonically.
+  const Envoy::MonotonicTime new_time = time_source.monotonicTime();
+  const uint64_t elapsed = start_ == Envoy::MonotonicTime::min() ? 0 : (new_time - start_).count();
+  start_ = new_time;
```

Comment:
Warning: Attempting to reason about concurrency...

Do you see any weird effects running with concurrency > 1 on the test server?

I was imagining a situation with 2 test server threads operating almost in lockstep, but 1ns apart. Suppose the requests come in evenly every 1ms. If the test server got stalled for a few milliseconds and the OS ended up buffering several packets, and then the test server threads started consuming them round robin, wouldn't the timing reports end up alternating 0, 1ns, 999999ns, 1ns, 999999ns, 1ns, ...?

This would actually represent the true response timing coming out of the test server, so the code in this PR would be working as intended.

Extreme statistics like that could give the impression that there is extreme distortion from the intermediary proxy, but it's entirely an artifact of the Nighthawk test server having two threads and stalling for a few milliseconds.

(Also the intermediary might also introduce its own equally extreme irregularities, since it could also be a multithreaded Envoy.)

The obvious way to avoid this issue is to use concurrency=1 on the test server. Then we would only be measuring the distortions from intermediaries and the OS.

Do you think there's any future reporting that could diagnose this issue? For example, if we also had exactly this PR but with thread-local stopwatches, we would see perfect evenness rather than 1-999999 fluctuation. We would know that the 1-999999 was the fault of our own multithreading, and any timing deviations would be our OS or the intermediary.

Reply:
Actually, maybe varying between 1ns and 1ms would be considered negligible distortion, especially if the intermediary added multiple milliseconds.

Reply:
Your reasoning here makes sense to me. The intent of this is to be able to sanity check part of our timings, so if it would catch those extremes where we don't expect them, I think it would be providing value.

The way I see it is that this is going to be particularly useful in A/B tests, for example:
- across different clients
- with and without an intermediary
- with nighthawk client features that might perform relatively heavy lifting on cpu cores shared with the workers (e.g. stats flushing).

Some existing OSS clients can probably easily be tweaked to take these as inputs to their latency reports to figure out their request-release timing precisions. 

Also, something like the delta between the expected and observed curves could perhaps serve as a means to quantify distortion; Indeed I think the extremes in the scenario you described would be signal in this case, and not noise (looking at you, adaptive load controller :-)).

As for tracking thread-local values: I've been thinking about that, and I think it's pretty doable doing that on the server side by adding a new Envoy histogram in the extension. On the client side however, I think it will take more consideration and work, as I think there would be some feature gaps:
- be able to set multiple response headers for tracking, or create a means to read multiple values from the header.
- if we want to be able to correlate thread id's we need to be able to tag values too, but this may not be nessecary.
- having a means to dynamically create histograms or a histogram capable of grouping on a tag would be needed
  for the Nighthawk client to handle it

Reply:
Afterthought: maybe, if we'd have the per worker histograms on the test server side, plus a means to read/sample these histograms on the NH client side and propagate these to the output, that might be a relatively low effort means to achieve this.

Reply:
Last thought: per-worker tracking may also yield data that's interesting in light of exposing unbalancedness.
I filed #487 to get this on the radar and track this / serve as a point for further discussion.